xarray/backends/core.py

```diff
@@ -0,0 +1,425 @@
+        store_variable = Variable(variable.dims,
+                                  np.empty_like(variable),
+                                  copy.deepcopy(variable.attrs))
+        self._variables[name] = store_variable
+        return store_variable.values
+
+    def get_writable_array(self, name):
+        return self._variables[name].values
+
+    def set_attribute(self, name, value):
+        self._attributes[name] = copy.deepcopy(value)
+
+    def get_write_lock(self, name, region=Ellipsis):
+        return self._write_locks[name]
+
+
+# Reading and writing with DataStores
+
+# This would live in xarray/backends/api.py, and would be user-facing as part
+# of the developer API. Note that it is entirely domain agnostic:
+
+def read_datastore(store, decode=None, chunks=None, close_on_error=False):
+
+    # open_dataset now simply opens an appropriate store, creates a default
+    # CFDecoder, and passes the arguments off to read_datastore.
+
+    if decode is None:
+        decode = coders.DummyDecoder()
+
+    try:
+        variables = store.get_variables()
+        attrs = store.get_attributes()
+        dataset = decode(variables, attrs)
+
+        if chunks is not None:
+            try:
+                from dask.base import tokenize
+            except ImportError:
+                import dask  # raise the usual error if dask is entirely missing
+
+            lock = {k: store.get_lock(k) for k in variables}
+            token = tokenize(store.get_token(), pickle.dumps(decoder))
+            prefix = store.get_name() + '/'
+            dataset2 = dataset.chunk(
+                chunks, name_prefix=prefix, token=token, lock=lock)
+            dataset2._file_obj = store
+        else:
+            dataset2 = dataset
+
+    except Exception:
+        # This exists for the sake of testing (especially on Windows, where
+        # unclosed files lead to errors)
+        if close_on_error:
+            store.close()
+
+    return dataset2
+
+
+class ArrayWriter(object):
+    def __init__(self):
+        self.sources = []
+        self.targets = []
+        self.locks = []
+
+    def add(self, source, target, lock=None):
+        if isinstance(source, dask_array_type):
+            self.sources.append(source)
+            self.targets.append(target)
+            self.locks.append(lock)
+        else:
+            target[...] = source
+
+    def sync(self):
+        if self.sources:
+            import dask.array as da
+            # TODO: dask.array.store needs to be able to accept a list of Lock
+            # objects.
+            da.store(self.sources, self.targets, lock=self.lock)
+
+
+# write_datastore also becomes developer facing API. Dataset.dump_to_store
+# and the various other machinery is deprecated.
+
+def write_datastore(dataset, store, encode=None, encoding=None,
+                    close_on_error=False):
+    # TODO: add compute keyword argument to allow for returning futures, like
+    # dask.array.store.
```

Comment:
I suspect that you don't want to deal with futures directly.  Instead you want to expose a dask.graph that the distributed client can collect and replace with futures on its own.